Introduce `swiftgen template doc`

AliSoftware opened the pull request:
* [x] I've started my branch from the `develop` branch (gitflow)
  * [x] And I've selected `develop` as the "base" branch for this Pull Request I'm about to create
* [x] I've added an entry in the `CHANGELOG.md` file to explain my changes and credit myself  
      _(or added `#trivial` to the PR title if I think it doesn't warrant a mention in the CHANGELOG)_
  * [x] Add the entry in the appropriate section (Bug Fix / New Feature / Breaking Change / Internal Change)
  * [x] Add a period and 2 spaces at the end of your short entry description
  * [x] Add links to your GitHub profile to credit yourself and to the related PR and issue after your description

---

This PR:

 - [aca2804a20afa43cf68f797d41b05e80cb646500] Renames occurrences of "subcommand" in the docs with "parser" or "parser command" as appropriate for consistency
- [cc56a41f61287b903a52afd250b85d85d6b9f4a1] Renames `swiftgen templates` subcommand to `swiftgen template` (singular) and deprecate the plural form of the subcommand, which will be removed in next major release
- [82ed9ddf4fed739b8eb1c3eeb223f34ea5fd74d8] Introduces the `swiftgen template doc [parser] [template]` to quickly open the documentation for templates on GitHub

SwiftGen-Eve commented:
<!--
  0 Errors
  0 Warnings
  0 Messages
  5 Markdowns
-->

Hey 👋 I'm Eve, the friendly bot watching over SwiftGen 🤖

Thanks a lot for your contribution!



---

Seems like everything is in order 👍 You did a good job here! 🤝

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AliSoftware commented on Sources/SwiftGen/main.swift:
```diff
@@ -23,21 +24,37 @@ let main = Group {
+    $0.addCommand("list", "list bundled and custom templates", TemplatesCLI.list)
+    $0.addCommand("which", "print path of a given named template", TemplatesCLI.which)
+    $0.addCommand("cat", "print content of a given named template", TemplatesCLI.cat)
+    $0.addCommand("doc", "open the documentation for templates on GitHub", TemplatesCLI.cat)
```
Note: I know I could have mutualised those 4 `$0.addCommand` lines with the ones in `$0.group("template")` on line 33, but I thought it was probably not worth creating a helper function to mutualise those 4 entries just so we undo it all in SwiftGen 7 when we remove the deprecated `templates` group entirely, so I thought copy/pasting those 4 lines would make it easier to just delete them later in 7.0)

AliSoftware commented on Sources/SwiftGen/Commander/TemplatesCommands.swift:
```diff
@@ -36,6 +36,40 @@ enum TemplatesCLI {
+    var path = "templates/"
+    if let subcommand = subcommand {
+      // If we have a subcommand argument, ensure that is one of the allowed ones
+      guard let parserCLI = ParserCLI.command(named: subcommand) else {
```
(thinking out loud): I wonder if `Commander` would allow us to write `Argument<ParserCLI?>` instead of `Argument<String?>` if we make `ParserCLI` conform to `ArgumentConvertible`… and if so maybe there are some other places and commands where we could then use this trick? 🤔  Meh maybe not worth it after all 🤷‍♂️

AliSoftware commented on Sources/SwiftGen/main.swift:
```diff
@@ -23,21 +24,37 @@ let main = Group {
+    $0.addCommand("doc", "open the documentation for templates on GitHub", TemplatesCLI.doc)
+  }
+
+  $0.group("run", "run individual parser commands without a config file") {
```
Is this the name we want for that subcommand? An alternate solution I thought about was to call that `parse`, so that it reads e.g. `swiftgen parse xcassets --templateName swift5 -o Assets.swift Images.xcassets`… thoughts?

djbe replied:
"Parse" sounds ok. If I'm not mistaken we *do* already have a `config run` command, so there is a precedent there.

AliSoftware commented:
Ok maybe I should have split that PR in two after all (`swiftgen template doc` vs `swiftgen run` + subcommand->parser rename) but realised it too late…

@djbe tell me if you think I need to split or if it's ok to keep it as one

[EDIT] I've finally split this into this current PR for `swiftgen template doc`, and #705 for `swiftgen run <parser>`, makes it all clearer

djbe commented on CHANGELOG.md:
```diff
@@ -13,12 +13,22 @@
   [David Jennes](https://github.com/djbe)
   [#614](https://github.com/SwiftGen/SwiftGen/pull/614)
+  
+### Deprecations
```
Note: if we're adding a "deprecations" section, we may want to check the other entries and see if they need to be moved here. For example, the entry right above this is one.

AliSoftware replied:
Good point.

AliSoftware replied:
Worth mentioning in #621 so we don't forget to reorder/revisit the CHANGELOG before releasing

djbe commented on CHANGELOG.md:
```diff
@@ -13,12 +13,22 @@
 * You can now easily create a new config file using `swiftgen config init`. This will create an example and commented config file and open it to let you edit it to your needs.  _Note that the generated config file is static content which doesn't take the user's project into account (though that might change in the future)_.  
   [@AliSoftware](https://github.com/AliSoftware)
   [#694](https://github.com/SwiftGen/SwiftGen/pull/694)
+* You can now use `swiftgen template doc [parser] [template]` on the command line to quickly open the documentation for templates on GitHub direclty from your terminal.  
```
Should that be `templateName`? Just to not confuse users that they can use it for custom templates. Maybe I'm worrying too much 😅

AliSoftware replied:
Yeah good point, worth being more specific.

```suggestion
* You can now use `swiftgen template doc [parser] [templateName]` on the command line to quickly open the documentation for templates on GitHub direclty from your terminal.  
```

NB: I think it's enough to be more specific in CHANGELOG only but not change the code though, as if you try to run `swiftgen template doc xcassets my-custom-template` then the implementation checks for bundled templates only (doesn't consider templates from Application Support nor custom templates) and exits with explicit error if template name not found.

AliSoftware commented:
Ok, I've now split the changes in separate PRs.

 - This one only focuses on `swiftgen template` (renaming `templates` to `template` + `template doc`, and also uniformise usage of `parser` instead of `subcommand` in APIs and docs
 - #705 focuses on `swiftgen run`

I've also split this PR #697 into clear commits if it makes the review easier, especially the renaming of `subcommand` to `parser` is separate from the deprecation of `templates` and from the introduction to `template doc` command. See PR description.

djbe commented on CHANGELOG.md:
```diff
@@ -13,12 +13,22 @@
 * You can now easily create a new config file using `swiftgen config init`. This will create an example and commented config file and open it to let you edit it to your needs.  _Note that the generated config file is static content which doesn't take the user's project into account (though that might change in the future)_.  
   [@AliSoftware](https://github.com/AliSoftware)
   [#694](https://github.com/SwiftGen/SwiftGen/pull/694)
+* You can now use `swiftgen template doc [parser] [templateName]` on the command line to quickly open the documentation for templates on GitHub direclty from your terminal.  
```
```suggestion
* You can now use `swiftgen template doc [parser] [templateName]` on the command line to quickly open the documentation for templates on GitHub directly from your terminal.  
```
